x-pack/plugins/monitoring/server/rules/base_rule.ts

```diff
@@ -309,13 +344,19 @@ export class BaseRule {
           nodeState.ui.message = this.getUiMessage(nodeState, node);
           // store the state of each node in array.
           newAlertStates.push(nodeState);
-        }
-        const alertInstanceState = { alertStates: newAlertStates };
-        // update the alert's state with the new node states
-        instance.replaceState(alertInstanceState);
-        if (newAlertStates.length) {
-          this.executeActions(instance, alertInstanceState, null, cluster);
-          state.lastExecutedAction = currentUTC;
+
+          const alertInstanceState = { alertStates: newAlertStates };
```

Comment:
I moved this to only report an alert if `node.shouldFire` is true. Previously, with the `alertFactory`, you could use `alertFactory.create` to create the alert but it wouldn't actually be considered an alert unless you called `scheduleActions` on it. With the `alertsClient`, it is a single call to `report` that both creates the alert and schedules actions for it.

Reply:
For my understanding, `report` will fire the alert while `setAlertData` executed later will cache the new state of the alert on the client ?

Reply:
The framework stores everything reported (using `report` and `setAlertData`) about the alert in memory until the rule executor is done, then it does a bunch of post-processing, so yes, `report` will fire the alert but this won't actually happen until the rule executor is finished. When you report an alert, you are able to get some info back about it (the UUID and the start time). The reason we added `setAlertData` was because there were some rules that wanted to use the UUID and/or start time inside its context/state variables.

Since these rules don't need that, we could technically move the `report` call into each rule type's `executeActions` function, so you'd call `alertsClient.report({id, actionGroup, context, state, payload})` but I was trying to minimize changes